public_html/wp-content/mu-plugins/blocks/assets/src/shared/featured-image/index.js

```diff
@@ -0,0 +1,183 @@
+	 * @param props.availableSizes Available sizes of image along with source URL. Not required if `props.wpMediaDetails` is present. Should be in this format:
+	 * [
+	 * 		{
+	 * 		 	height: 999,
+	 * 		 	width: 999,
+	 * 		 	source_url: https://link.to/image-size-1.png
+	 * 		},
+	 * 		{
+	 * 			height: 500,
+	 * 			width: 500,
+	 * 			source_url: https://link.to/image-size-2.png
+	 * 		}
+	 * 		...
+	 * ]
+	 * @param props.height Height in pixels for image.
+	 * @param props.width Width in pixels for image.
+	 * @param props.className Classname for image element
+	 * @param props.alt Alt text for image
+	 */
+	constructor( props ) {
+		super( props );
+	}
+
+	componentWillMount() {
+		let availableSizes;
+		const { wpMediaDetails } = this.props;
+		if ( wpMediaDetails ) {
+			availableSizes = FeaturedImage.parseWPMediaDetails( wpMediaDetails );
+		} else {
+			availableSizes = this.props.availableSizes;
+		}
+		this.parseAvailableSizes( availableSizes );
+	}
+
+	/**
+	 * Converts from WP API featured_media object to array of different sizes.
+	 *
+	 * @param wpMediaDetails
+	 */
+	static parseWPMediaDetails( wpMediaDetails ) {
+		const availableSizes = [];
+		for ( const size of Object.keys( wpMediaDetails ) ) {
+			if ( ! wpMediaDetails.hasOwnProperty( size  ) ) {
+				continue;
+			}
+
+			availableSizes.push( wpMediaDetails[ size ] );
+		}
+		return availableSizes;
+	}
+
+	/**
+	 * Pre-calculate and store aspect ratio of all images. This will help us in picking the best image available for the size that we have.
+	 */
+	parseAvailableSizes( availableSizes ) {
+		let largestSizeValue = 0;
+		let largestSize = {};
+		// Lets pre-calculate aspect ratio of all images. Also cache the largest area image.
+		const sizes = availableSizes.map( ( size ) => {
+			return {
+				aspectRatio: size.width / size.height,
+				area: size.width * size.height,
+				...size
+			}
+		} );
+
+		// Lets sort the available images based on area and then aspect ratio.
+		sizes.sort( ( size1, size2 ) => {
+				if ( size1.area < size2.area ) {
+					return -1;
+				} else if ( size1.area === size2.area ) {
+					return size1.aspectRatio < size2.aspectRatio ? -1 : 1;
+				} else {
+					return 1;
+				}
+			}
+		);
+
+		this.setState(
+			{
+				sizes,
+				...this.state
+			}
+		);
+	}
+
+
+	/**
+	 * Returns URL of appropriate image. We find appropriate by looking for an image with similar aspect ratio, and similar or greater area.
+	 * If there is no image with similar aspect ratio or area, we will return strictly bigger image then currently requested.
+	 * If there is no bigger image then currently request one, we will return the max size image URL.
+	 * Note: Core Image block allows user to select image size by name, but we cannot really do that because we will be controlling multiple images at once, and one or more images may not have all the available sizes. So an automated solution is preferred here.
```

Comment:
Kudos, I think this sounds like a good approach. 

> we will be controlling multiple images at once, and one or more images may not have all the available sizes.

I wonder, though, if it would be worth still letting the user pick a size, and then use these algorithms as a backup for the case when the correct size isn't available. That way, when the size _is_ available, we could skip all these calculations...

Reply:
🤔 thats a good point. So in this case this component will need two inputs, one would be of size name, and other would be an object of `{ height, width }`. We will use height, width and this algo if size is not available.

Reply:
👍

Reply:
So I ended with not having pre-defined sizes, but having height and width pass via props. Algo will automatically pick exact size if its available, otherwise will make an effort to find appropriately sized image.

Reply:
Tracking this in https://github.com/WordPress/wordcamp.org/issues/36. Resolving this conversation.